homeassistant/components/husqvarna_automower/switch.py

```diff
@@ -0,0 +1,69 @@
+"""Creates a switch entity for the mower."""
+import logging
+
+from aioautomower.exceptions import ApiException
+from aioautomower.model import MowerStates
+
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import AutomowerDataUpdateCoordinator
+from .entity import AutomowerBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up switch platform."""
+    coordinator: AutomowerDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        AutomowerSwitchEntity(mower_id, coordinator) for mower_id in coordinator.data
+    )
+
+
+class AutomowerSwitchEntity(AutomowerBaseEntity, SwitchEntity):
+    """Defining the Automower switch."""
+
+    _attr_translation_key = "park_until_further_notice"
+
+    def __init__(
+        self,
+        mower_id: str,
+        coordinator: AutomowerDataUpdateCoordinator,
+    ) -> None:
+        """Set up Automower switch."""
+        super().__init__(mower_id, coordinator)
+        self._attr_unique_id = f"{self.mower_id}_park_until_further_notice"
+
+    @property
+    def is_on(self) -> bool:
+        """Return the state of the switch."""
+        attributes = self.mower_attributes
+        return (
+            attributes.mower.state == MowerStates.RESTRICTED
+            and attributes.planner.restricted_reason == "NOT_APPLICABLE"
+        )
+
+    async def async_turn_on(self, **kwargs):
+        """Turn the entity on."""
+        try:
+            await self.coordinator.api.park_until_further_notice(self.mower_id)
```

Comment:
Ok, so the switch in on-state means the mower will park itself and wait for the toggle to off to start mowing again?

Wouldn't it be better to implement a service for such a feature? Together with a binary_sensor to show the state: parked/mowing?

I can see the reason for using a switch but I wonder is this following the Dev rules?

Reply:
> Ok, so the switch in on-state means the mower will park itself and wait for the toggle to off to start mowing again?

Yes
> Wouldn't it be better to implement a service for such a feature? Together with a binary_sensor to show the state: parked/mowing?

A button and a binary sensor would be an alternative, yes. But I think this gives the best user experience.

> I can see the reason for using a switch but I wonder is this following the Dev rules?
Not sure...

Reply:
After checking the HA Developer Docs, I think a Select would be a better fit to what you want to achieve: https://developers.home-assistant.io/docs/core/entity/select

The Switch platform is meant for turning something on or off. That's not what your switch is doing.

Selecting between the options `mowing the lawn` and `parked until further notice` makes more sense.

Reply:
Wait, there is this which is the perfect fit :smile: https://developers.home-assistant.io/docs/core/entity/lawn-mower

Reply:
In the select you can connect the action to execute to a select-able option.

Reply:
See for reference "my" work: https://github.com/home-assistant/core/blob/dev/homeassistant/components/plugwise/select.py

Reply:
What would the options be for a select entity in this case? Could we detect which option would be in use from the state of the mower device? Can we extend the select entity with more options and still detect what option is in use?

Reply:
The options could be: `start_mowing`, `pauze` and `dock`, implementing access to the three "set"-functions in `lawnmover.py`.

Reply:
Why would we do that? They are already available in the lawn mover entity.

Reply:
I'm thinking @Thomas55555 wants to provide easy access to the lawn_mower set-functions/services?
I'm thinking there is no lawn_mower card that gives the user access to the set-functions? Or is there?

Reply:
My interpretation of this PR is to give access to the setting to enable and disable the schedule as described in the architecture discussion:
https://github.com/home-assistant/architecture/discussions/894#discussioncomment-5873265

Reply:
Ah OK, that is the background of this PR, thanks for providing the info :)

"should be represented by a separate switch entity per schedule": if there can be more schedules than one, a select with an added off-option would be more suitable?

Looking at the code again, I would suggest switching the on/off implementations: turn_on should enable the schedule, turn_off should result to parking.

But, will this really achieve what is requested/intended? Does a schedule not include parking-time?

Reply:
I see that there is also a `park_until_next_schedule` option in the backend, so yes, the `park_until_further_notice` option does disable the active schedule.

Reply:
Summarizing: if there is a single calendar with a schedule, then the switch is the correct implementation.
My feedback for this is: switch the turn_on/_off options: turn_on should enable the schedule (`resume_schedule`).

If there can exist more than one schedule then I would suggest implementing a Select.

Reply:
Why is a select better than a switch per schedule?

Reply:
When there are more than one switch the user can enable more than one Switch. That's not possible with a Select.
See the Plugwise climate thermostat_schedule select for reference:
![image](https://github.com/home-assistant/core/assets/11290930/5fd0434b-f6d7-482b-a8d8-32b3b16729f5)

Reply:
Looking at the backend code, and at the Husqvarna webpage, my impression is that there is only one Schedule so implementing a single Switch would be OK.

Reply:
Yes, if there is more than one schedule and only one schedule can be active at the same time, a select entity is better than a switch.

Reply:
@Thomas55555 Does the lawnmower App provide a single schedule/calendar?

Reply:
There are multiple schedules possible in the app. But you cannot select to use just a specific schedule. Either you use all of them or not.
The sentence above is valid for **all** mowers.
There are **some** mowers, which support workareas. There you can have separate schedules for each work area. But you can't turn them of, or select a schedule. So in my opinion the switch is the best option.

Reply:
With multiple schedules you mean for instance different schedules for different days? Or weeks or months? In winter you mow less than in summer, is that what you mean?

Reply:
In principle that is still one schedule, you use the term calendar I think.

With such a single calendar, a single switch is OK for me as well.
But please make the function as I wrote above: switch the functions of turn_on/_off and change the name so that it's clear that the schedule is active when the switch is on.

Reply:
> With multiple schedules you mean for instance different schedules for different days? Or weeks or months? In winter you mow less than in summer, is that what you mean?

I mean, multiple schedules can be:
Schedule 1: monday 9:00 - 11:00
Schedule 2: thursday 7:00 - 10:00

Reply:
It's possible. But the better way is to create an calendar entity I think. There it would also be possible to change, delete and create the schedule-entries.
I've already did that in the custom component:
https://github.com/Thomas55555/husqvarna_automower/blob/main/custom_components/husqvarna_automower/calendar.py

Reply:
Ok, it's possible, that sound complicated.

Anyway, if there's a (single) calendar, better to create a single switch that allows the user to block the execution of upcoming calendar-tasks.

Reply:
> Ok, it's possible, that sound complicated.
> 
> Anyway, if there's a (single) calendar, better to create a single switch that allows the user to block the execution of upcoming calendar-tasks.

I thought about how to implement that... The problem is, if we do that, we have to save the deactivated schedule in the config entry, because it's not possible to disable the schedule over the API. We just can delete schedules in the API, not disable them.

Reply:
You're thinking the wrong way now, that was not my intention :)

As I see it, your original intention was to implement a switch, when turned on, parks to mower until further notice, effectively disabling the execution of the schedule-items in the calendar.

I'm proposing that you modify this idea somewhat, my latest idea is that you make it so that the switch, when turned on, pauses the execution of the schedule-items in the calendar. Your code already does that :smile: But you should change the name to something like: "Pause calendar". Because that's what the user wants to achieve: pause the mowing according to the calendar.

Reply:
It should be in the state `parked_until_further_notice`.

Reply:
Then why not just call it like that? The user knows this command from the Android/iOS app. Pause calendar is confusing i think. This could also mean "mow forever".

Reply:
We're getting back to my original point: I don't think a switch on-/off-state should be called `parked_until_further_notice`.
A switch-state should turn something on or off, or with some more liberty, enable or disable something.
What could be that "something" according to you? This "something" must be the name of the switch.

Reply:
The switch was called `park_until_further_notice`.  Without "ed". Is that maybe where the confusion comes from, or should in be named in `park_until_further_notice_mode`?

Reply:
To be honest, if so many words are needed to describe the function, it's not a good description for a switch-function. 

I'm wondering, why does a user put the mower in the `park_until_further_notice_mode`?
Ok, I went back to the backend, I see 6 options:
- resume_schedule
- pause_mowing
- park_until_next_schedule
- park_until_further_notice
- park_for (duration)
- start_for (duration)

With this info in front of me, a single switch does not make sense anymore.
I'm back at the Select-idea: create a Select that allows you to choose between the first 4 options. And you can keep the names, the Select gives you that freedom.
All 6 options would be better, but how to let the user enter a duration? Not supported I think. @MartinHjelmare?

Reply:
What's wrong with the code as it is now?

Reply:
The newest code is fine by me :)